document examples (#8)
JuliaGaussianProcesses/JuliaGPsDocs.jl: docs/examples.md

# Example notebooks

In JuliaGaussianProcesses, each notebook example is stored in its own subdirectory `<package-root>/examples/<example-dir>`.
The notebook itself is stored in [Literate.jl](https://github.com/fredrikekre/Literate.jl) format in a file called `script.jl`.

To run them locally, navigate to the directory with the example that you want to run and
start the Julia REPL and activate the project environment of the example:
```julia

> julia> ] activate .

Reviewer: I would generally recommend to go to to the package root and run `] activate examples/example-a` but that's my workflow.

```
Alternatively, you can start Julia with `julia --project=.`. Then install all required
packages with
```julia
julia> ] instantiate
```
Afterwards simply run
```julia
julia> include("script.jl")
```
In particular when editing an example, it can be convenient to (re-)run only some parts of
an example.
Many editors with Julia support such as VSCode, Juno, and Emacs support the evaluation of individual lines or code chunks.

You can convert a notebook to markdown and Jupyter notebook formats, respectively, by executing
```julia
julia> using Literate
julia> Literate.markdown("script.jl", "output_directory")
julia> Literate.notebook("script.jl", "output_directory")
```
(see the [Literate.jl docs](https://fredrikekre.github.io/Literate.jl/v2/) for additional options) or run
```shell
julia docs/literate.jl myexample output_directory
```
which also executes the code and generates embedded plots etc. in the same way as in building the AbstractGPs documentation.

## Add a new example

> Create a new subdirectory in `<package-root>/examples`, and in that

Reviewer: Maybe directly give an example like `<package-root>/examples/my-example`

Author: ```suggestion
Create a new subdirectory `<package-root>/examples/<new-example>`, and
```

> subdirectory put your code in a file called `script.jl` so that it will get

Author: ```suggestion
put your code in the file `<package-root>/examples/<new-example>/script.jl` so that it will get
```

picked up by the automatic docs build.

Every example uses a separate project environment. Therefore you should also create a new
project environment in the directory of the example that contains all packages required by your script.
Note that the dependencies of your example *must* include the `Literate` package.

From a Julia REPL started in your example script's directory, you can run
```julia
julia> ] activate .

> julia> ] add Literate

Reviewer: This should not be necessary thanks to David's trick.

Author: mhm I think currently it's still required? see https://github.com/JuliaGaussianProcesses/JuliaGPsDocs.jl/blob/main/test/examples/example-a/Project.toml

we can update it when it actually gets implemented here

Author: Update it once https://github.com/JuliaGaussianProcesses/JuliaGPsDocs.jl/issues/2 is resolved

# Example notebooks

> In JuliaGaussianProcesses, each notebook example is stored in its own subdirectory `<package-root>/examples/<example-dir>`.

Author: ```suggestion
In JuliaGaussianProcesses, each notebook example is stored in its own subdirectory `<PACKAGE-ROOT>/examples/<EXAMPLE-SUBDIR>`.
```

The notebook itself is stored in [Literate.jl](https://github.com/fredrikekre/Literate.jl) format in a file called `script.jl`.

> To run them locally, navigate to the directory with the example that you want to run and
> start the Julia REPL and activate the project environment of the example:
> ```julia
> julia> ] activate .

Reviewer: ```suggestion
To run them locally, navigate to the package directory, start the Julia REPL
and activate the environment of the example with the correct path:
```julia
julia> ] activate examples/my-example
```

> To run them locally, navigate to the directory with the example that you want to run and
> start the Julia REPL and activate the project environment of the example:
> ```julia
> julia> ] activate .
> ```

Reviewer: ```suggestion
To run them locally, navigate to the package directory, start the Julia REPL
and activate the environment of the example with the correct path:
```julia
julia> ] activate examples/my-example
```
```

Reviewer: Looks like it cut my suggestion...

> Alternatively, you can start Julia with `julia --project=.`. Then install all required

Author: ```suggestion
Alternatively, you can start Julia with `julia --project=examples/my-example`. Then install all required
```

packages with
```julia
julia> ] instantiate
```
Afterwards simply run
```julia

> julia> include("script.jl")

Reviewer: ```suggestion
julia> include("examples/my-example/script.jl")
```

```
In particular when editing an example, it can be convenient to (re-)run only some parts of
an example.
Many editors with Julia support such as VSCode, Juno, and Emacs support the evaluation of individual lines or code chunks.

You can convert a notebook to markdown and Jupyter notebook formats, respectively, by executing
```julia
julia> using Literate

> julia> Literate.markdown("script.jl", "output_directory")

Author: ```suggestion
julia> Literate.markdown("examples/my-example/script.jl", "output_directory")
```

> julia> Literate.notebook("script.jl", "output_directory")

Author: ```suggestion
julia> Literate.notebook("examples/my-example/script.jl", "output_directory")
```

```
(see the [Literate.jl docs](https://fredrikekre.github.io/Literate.jl/v2/) for additional options) or run
```shell
julia docs/literate.jl myexample output_directory
```

> which also executes the code and generates embedded plots etc. in the same way as in building the AbstractGPs documentation.

Author: Need to update this!

> (see the [Literate.jl docs](https://fredrikekre.github.io/Literate.jl/v2/) for additional options) or run
> ```shell
> julia docs/literate.jl myexample output_directory
> ```
> which also executes the code and generates embedded plots etc. in the same way as in building the AbstractGPs documentation.

Author: ``````suggestion
(see the [Literate.jl docs](https://fredrikekre.github.io/Literate.jl/v2/) for additional options).

To execute the code and generate embedded plots etc. in the same way as in the official building of the JuliaGaussianProcesses documentation, you can run the following (assuming you are in the package root directory):
```julia
julia> ] activate --temp
julia> using Pkg
julia> Pkg.add(Pkg.PackageSpec(; url="https://github.com/JuliaGaussianProcesses/JuliaGPsDocs.jl")) # While the package is unregistered, it's a workaround
julia> using JuliaGPsDocs
julia> JuliaGPsDocs.run_examples(["examples/my-example"], "my-output-dir", "examples", ".", "")
```
``````

## Add a new example

Create a new subdirectory `<package-root>/examples/<new-example>`, and
put your code in the file `<package-root>/examples/<new-example>/script.jl` so that it will get
picked up by the automatic docs build.

Every example uses a separate project environment. Therefore you should also create a new
project environment in the directory of the example that contains all packages required by your script.

> Note that the dependencies of your example *must* include the `Literate` package.

Author: ```suggestion
Note that the dependencies of your example *must* include the `Literate` package (see https://github.com/JuliaGaussianProcesses/JuliaGPsDocs.jl/issues/2).
```

> From a Julia REPL started in your example script's directory, you can run
> ```julia
> julia> ] activate .
> julia> ] add Literate
> julia> # the following line adds an AbstractGPs dependency that is based on the local directories, not a hash:
> julia> ] dev ../..
> julia> # add any other example-specific dependencies

Reviewer: ```suggestion
From a Julia REPL started in your package directory, you can run
```julia
julia> ] activate examples/my-example
julia> ] add Literate
julia> # the following line adds an AbstractGPs dependency that is based on the local directories, not a hash:
julia> ] dev .
julia> # add any other example-specific dependencies
```
